### ISSUE TITLE
Improve error line numbers in doctests

### ISSUE THREAD

**Issue by mystor** [rust-lang/rust#45556]
Currently if an error (such as a build error) occurs within a doctest, the line numbers are relative to the doc comment's start point, rather than actual lines in the file, which is annoying.

I think that it would be better if the errors were reported based on lines in the original source file.

**Comment by kennytm** [comment]
The problem of fixing it is that, it may break doc tests that uses `line!()`, unless we decouple the line number reported by `line!()` and the line number reported by the error diagnostic.

**Comment by mystor** [comment]
Would we be able to crater this & see if it actually affects people's doctests in the wild?

**Comment by Mark-Simulacrum** [comment]
Error today; seems like this has been fixed.

```
---- src/lib.rs - foo (line 3) stdout ----
error[E0369]: binary operation `+` cannot be applied to type `&str`
 --> src/lib.rs:5:18
  |
3 |     let _x = "t" + 10;
  |              --- ^ -- {integer}
  |              |
  |              &str
  |
  = note: an implementation of `std::ops::Add` might be missing for `&str`

error: aborting due to previous error
```
